experimental/dds/tree2/src/feature-libraries/chunked-forest/codec/chunkDecoding.ts

```diff
@@ -0,0 +1,257 @@
+		assert(chunk.subChunks.length > 0, "Unexpected empty sequence");
+		// Logic in this file is designed to never produce an unneeded (single item) SequenceChunks, so its better to throw an error here to detect bugs.
+		assert(chunk.subChunks.length > 1, "Unexpected single item sequence");
+
+		for (const sub of chunk.subChunks) {
+			// The logic in this file is designed to never produce an nested SequenceChunks or emptyChunk, so its better to throw an error here to detect bugs.
+			assert(!(sub instanceof SequenceChunk), "unexpected nested sequence");
+			assert(sub !== emptyChunk, "unexpected empty chunk");
+
+			sub.referenceAdded();
+		}
+
+		chunk.referenceRemoved();
+		return chunk.subChunks;
+	} else {
+		return [chunk];
+	}
+}
+
+export function aggregateChunks(input: TreeChunk[]): TreeChunk {
+	const chunks = input.flatMap(deaggregateChunks);
+	switch (chunks.length) {
+		case 0:
+			return emptyChunk;
+		case 1:
+			return chunks[0];
+		default:
+			return new SequenceChunk(chunks);
+	}
+}
+
+export class NestedArrayDecoder implements ChunkDecoder {
+	public constructor(private readonly shape: EncodedNestedArray) {}
+	public decode(decoders: readonly ChunkDecoder[], stream: StreamCursor): TreeChunk {
+		const decoder = decoders[this.shape];
+
+		// TODO: uniform chunk fast path
+		const chunks: TreeChunk[] = [];
+
+		const data = readStream(stream);
+		if (typeof data === "number") {
+			// This case means that the array contained 0 sized items, and was thus encoded as the length instead of the array.
+			const inner = { data: [], offset: 0 };
+			for (let index = 0; index < data; index++) {
+				chunks.push(decoder.decode(decoders, inner));
+			}
+		} else {
+			assert(Array.isArray(data), "expected number of array for encoding of nested array");
+			const inner = { data, offset: 0 };
+			while (inner.offset !== inner.data.length) {
+				chunks.push(decoder.decode(decoders, inner));
+			}
+		}
+
+		return aggregateChunks(chunks);
+	}
+}
+
+export class InlineArrayDecoder implements ChunkDecoder {
+	public constructor(private readonly shape: EncodedInlineArray) {}
+	public decode(decoders: readonly ChunkDecoder[], stream: StreamCursor): TreeChunk {
+		const length = this.shape.length;
+		const decoder = decoders[this.shape.shape];
+		const chunks: TreeChunk[] = [];
+		for (let index = 0; index < length; index++) {
+			chunks.push(decoder.decode(decoders, stream));
+		}
+		return aggregateChunks(chunks);
+	}
+}
+
+export class AnyDecoder implements ChunkDecoder {
+	public static readonly instance = new AnyDecoder();
+	private constructor() {}
+	public decode(decoders: readonly ChunkDecoder[], stream: StreamCursor): TreeChunk {
+		const shapeIndex = readStreamNumber(stream);
+		const decoder = getChecked(decoders, shapeIndex);
+		return decoder.decode(decoders, stream);
+	}
+}
```

Comment:
Any reason to use a singleton class here over something like this?


```suggestion
export const anyDecoder: ChunkDecoder = {
	decode(decoders: readonly ChunkDecoder[], stream: StreamCursor): TreeChunk {
		const shapeIndex = readStreamNumber(stream);
		const decoder = getChecked(decoders, shapeIndex);
		return decoder.decode(decoders, stream);
	},
};
```

Reply:
Reasons to do it this way are classes are more readable in the debugger, and its more consistent with other implementations.

That said, I still prefer your version, so I'll change it.

Reply:
Done.